File: leetcode/535.tiny-url-的加密与解密.cpp
#include <iostream>
#include <unordered_map>
using namespace std;
/*
 * @lc app=leetcode.cn id=535 lang=cpp
 *
 * [535] TinyURL 的加密与解密
 */

// @lc code=start
class Solution
{
private:
    unordered_map<int, string> data;
    int max_id = 0;

public:
    // Encodes a URL to a shortened URL.
    string encode(string longUrl)
    {
        max_id++;
        data[max_id] = longUrl;
        return "http://tinyurl.com/" + to_string(max_id);
    }

    // Decodes a shortened URL to its original URL.
    string decode(string shortUrl)
    {
        string data_id = shortUrl;
        size_t pos;
        string delimiter = "/";
        while ((pos = data_id.find(delimiter)) != string::npos)
        {
            data_id.erase(0, pos + delimiter.length());
        }
        return data[std::stoi(data_id)];
    }
};

// @lc code=end
int main()
{
    string url = "https://leetcode.com/problems/design-tinyurl";
    Solution solution;
    return url != solution.decode(solution.encode(url));
}

File: leetcode/535.tiny-url-的加密与解密.py
#
# @lc app=leetcode.cn id=535 lang=python3
#
# [535] TinyURL 的加密与解密
#

# @lc code=start
class Codec:

    def __init__(self) -> None:
        self.data = {}
        self.max_id = 0

    def encode(self, longUrl: str) -> str:
        """Encodes a URL to a shortened URL.
        """
        self.max_id += 1
        self.data[self.max_id] = longUrl
        return f"http://tinyurl.com/{self.max_id}"

    def decode(self, shortUrl: str) -> str:
        """Decodes a shortened URL to its original URL.
        """
        data_id = shortUrl.split("/")[-1]
        return self.data.get(int(data_id), "")


# @lc code=end
# Your Codec object will be instantiated and called as such:
codec = Codec()
assert "https://leetcode.com/problems/design-tinyurl" == codec.decode(
    codec.encode("https://leetcode.com/problems/design-tinyurl")
)
